clippy_lints/src/unnecessary_self_imports.rs

```diff
@@ -0,0 +1,177 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::in_macro;
+use clippy_utils::source::snippet_with_applicability;
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::{Item, ItemKind, Mod, Node, Path, UseKind, VisibilityKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::def_id::DefId;
+use rustc_span::symbol::Ident;
+use rustc_span::{BytePos, Span};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for imports ending in `::{self};`.
+    ///
+    /// **Why is this bad?** In most cases, this can be written much more cleanly by omitting `self`.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// use std::io::{self};
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::io;
+    /// ```
+    pub UNNECESSARY_SELF_IMPORTS,
+    style,
+    "imports ending in `self`, which can be omitted"
+}
+
+declare_lint_pass!(UnnecessarySelfImports => [UNNECESSARY_SELF_IMPORTS]);
+
+impl<'tcx> LateLintPass<'tcx> for UnnecessarySelfImports {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'_>) {
+        if_chain! {
+            if !in_macro(item.span);
+            if let ItemKind::Use(path, UseKind::ListStem) = &item.kind;
+            if let Some(ident) = is_self_import(cx, item, path);
+            if let Some(Res::Def(DefKind::Mod, def_id)) = path.segments[path.segments.len() - 2].res;
```

Comment:
This probably can't panic. But I would use pattern matching here either way.

```suggestion
            if let &[.., module, _self] = &path.segments;
            if let Some(Res::Def(DefKind::Mod, def_id)) = module.res;
```